npm/pkg/dataplane/ipsets/ipsetmanager.go

```diff
@@ -23,261 +25,346 @@ func (iMgr *IPSetManager) exists(name string) bool {
+	if err := iMgr.checkForListMemberUpdateErrors(listName, setNames, npmerrors.AppendIPSet); err != nil {
+		return err
+	}
 
-		// Nested IPSets are only supported for windows
-		// Check if we want to actually use that support
-		if set.Kind != HashSet {
-			return npmerrors.Errorf(npmerrors.DeleteIPSet, false, fmt.Sprintf("member ipset %s is not a Set type and nestetd ipsets are not supported", setName))
+	list := iMgr.setMap[listName]
+	for _, setName := range setNames {
+		if list.hasMember(setName) {
+			return npmerrors.Errorf(npmerrors.AppendIPSet, false, fmt.Sprintf("ipset %s is already a member of ipset %s", setName, listName))
 		}
+	}
+	return nil
+}
 
-		list, exists := iMgr.setMap[listName] // check if the Set exists
-		if !exists {
-			return npmerrors.Errorf(npmerrors.DeleteIPSet, false, fmt.Sprintf("ipset %s does not exist", listName))
+func (iMgr *IPSetManager) checkForRemoveFromListErrors(listName string, setNames []string) error {
+	if err := iMgr.checkForListMemberUpdateErrors(listName, setNames, npmerrors.DeleteIPSet); err != nil {
+		return err
+	}
+
+	list := iMgr.setMap[listName]
+	for _, setName := range setNames {
+		if !list.hasMember(setName) {
+			return npmerrors.Errorf(npmerrors.DeleteIPSet, false, fmt.Sprintf("ipset %s is not a member of ipset %s", setName, listName))
 		}
+	}
+	return nil
+}
 
-		if list.Kind != ListSet {
-			return npmerrors.Errorf(npmerrors.DeleteIPSet, false, fmt.Sprintf("ipset %s is not a list set", listName))
+func (iMgr *IPSetManager) checkForListMemberUpdateErrors(listName string, memberNames []string, npmErrorString string) error {
+	if !iMgr.exists(listName) {
+		return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s does not exist", listName))
+	}
+
+	list := iMgr.setMap[listName]
+	if list.Kind != ListSet {
+		return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s is not a list set", listName))
+	}
+
+	for _, memberName := range memberNames {
+		if listName == memberName {
+			return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s cannot be added to itself", listName))
+		}
+		if !iMgr.exists(memberName) {
+			return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s does not exist", memberName))
 		}
+		member := iMgr.setMap[memberName]
 
-		// check if Set is a member of List
-		_, exists = list.MemberIPSets[setName]
-		if !exists {
-			return nil
+		// Nested IPSets are only supported for windows
+		// Check if we want to actually use that support
+		if member.Kind != HashSet {
+			return npmerrors.Errorf(npmErrorString, false, fmt.Sprintf("ipset %s is not a hash set and nested list sets are not supported", memberName))
 		}
+	}
+	return nil
+}
 
-		// delete IPSet from the list
-		delete(list.MemberIPSets, setName)
-		set.DecIpsetReferCount()
-		// Update metrics of the IpSet
-		metrics.RemoveEntryFromIPSet(listName)
+func (iMgr *IPSetManager) addMemberIPSet(listName, memberName string) {
+	list := iMgr.setMap[listName]
```

Comment:
can we send in a pointer of the list obj, that way we are using up memory and waiting for GC to  clean it up ?

Reply:
this change would help 2 functions and for the other 2 it would use up memory i.e. add in the code `set := iMgr.setMap[setName]`

Left the change out for now to simplify changes

Reply:
addMemberIPSet this function seems to be used only in one place, so we should be able to send in the pointer right ?

Reply:
oh I was actually looking at `modifyCacheForKernelMemberUpdate()`

this and the function referenced in the comment below are called from functions that don't have the pointers, so we would need to add the code `set := iMgr.setMap[setName]` to them, which adds more for GC to clean up?

Reply:
that would be good too, since we are calling that function multiple times, a set obj pointer would need to be created multiple times we can get away with not duplicating by passing pointer

Reply:
we are going to remove the for loop in a followup PR though?